src/zon.zig

```diff
@@ -0,0 +1,661 @@
+    const tree = self.file.tree;
+    assert(tree.errors.len > 0);
+
+    const gpa = self.sema.gpa;
+    const ip = &self.sema.pt.zcu.intern_pool;
+    const parse_err = tree.errors[0];
+
+    var buf: std.ArrayListUnmanaged(u8) = .{};
+    defer buf.deinit(gpa);
+
+    // Create the main error
+    buf.clearRetainingCapacity();
+    try tree.renderError(parse_err, buf.writer(gpa));
+    const err_msg = try Zcu.ErrorMsg.create(
+        gpa,
+        .{
+            .base_node_inst = try ip.trackZir(gpa, .main, .{
+                .file = self.file_index,
+                .inst = .main_struct_inst,
+            }),
+            .offset = .{ .token_abs = parse_err.token + @intFromBool(parse_err.token_is_prev) },
+        },
+        "{s}",
+        .{buf.items},
+    );
+
+    // Check for invalid bytes
+    const token_starts = tree.tokens.items(.start);
+    const token_tags = tree.tokens.items(.tag);
+    if (token_tags[parse_err.token + @intFromBool(parse_err.token_is_prev)] == .invalid) {
+        const bad_off: u32 = @intCast(tree.tokenSlice(parse_err.token + @intFromBool(parse_err.token_is_prev)).len);
+        const byte_abs = token_starts[parse_err.token + @intFromBool(parse_err.token_is_prev)] + bad_off;
+        try self.sema.pt.zcu.errNote(
+            .{
+                .base_node_inst = try ip.trackZir(gpa, .main, .{
+                    .file = self.file_index,
+                    .inst = .main_struct_inst,
+                }),
+                .offset = .{ .byte_abs = byte_abs },
+            },
+            err_msg,
+            "invalid byte: '{'}'",
+            .{std.zig.fmtEscapes(tree.source[byte_abs..][0..1])},
+        );
+    }
+
+    // Create the notes
+    for (tree.errors[1..]) |note| {
+        if (!note.is_note) break;
+
+        buf.clearRetainingCapacity();
+        try tree.renderError(note, buf.writer(gpa));
+        try self.sema.pt.zcu.errNote(
+            .{
+                .base_node_inst = try ip.trackZir(gpa, .main, .{
+                    .file = self.file_index,
+                    .inst = .main_struct_inst,
+                }),
+                .offset = .{ .token_abs = note.token + @intFromBool(note.token_is_prev) },
+            },
+            err_msg,
+            "{s}",
+            .{buf.items},
+        );
+    }
+
+    try self.sema.pt.zcu.failed_files.putNoClobber(gpa, self.file, err_msg);
+    return error.AnalysisFail;
+}
+
+const Ident = struct {
+    bytes: []const u8,
+    owned: bool,
+
+    fn deinit(self: *Ident, allocator: Allocator) void {
+        if (self.owned) {
+            allocator.free(self.bytes);
+        }
+        self.* = undefined;
+    }
+};
+
+fn ident(self: LowerZon, token: Ast.TokenIndex) !Ident {
+    var bytes = self.file.tree.tokenSlice(token);
+
+    if (bytes[0] == '@' and bytes[1] == '"') {
+        const gpa = self.sema.gpa;
+
+        const raw_string = bytes[1..bytes.len];
+        var parsed = std.ArrayListUnmanaged(u8){};
```

Comment:
style nit:
```suggestion
        var parsed: std.ArrayListUnmanaged(u8) = .{};
```

Reply:
Also, using `gpa` here is questionable -- we want to minimize direct usage of gpa in Sema to avoid contention. This should probably be allocated into the Sema arena.